extensions/interactions/RatioExpressionInput/directives/ratio-expression-input-validation.service.ts

```diff
@@ -0,0 +1,200 @@
+        }
+      ];
+    } else if (!isInt(minimumNumberOfTerms)) {
+      return [
+        {
+          type: AppConstants.WARNING_TYPES.ERROR,
+          message: (
+            'Number of terms must be integral.')
+        }
+      ];
+    } else {
+      return [];
+    }
+  }
+
+  getAllWarnings(
+      stateName: string,
+      customizationArgs: RatioExpressionInputCustomizationArgs,
+      answerGroups: AnswerGroup[], defaultOutcome: Outcome): Warning[] {
+    let warningsList = [];
+    let ratioRulesService = (
+      new RatioExpressionInputRulesService(this.rof));
+    var minimumNumberOfTerms =
+      customizationArgs.numberOfTerms.value;
+    warningsList = warningsList.concat(
+      this.getCustomizationArgsWarnings(customizationArgs));
+
+    warningsList = warningsList.concat(
+      this.baseInteractionValidationServiceInstance.getAllOutcomeWarnings(
+        answerGroups, defaultOutcome, stateName));
+
+    // This validations ensures that there are no redundant rules present in the
+    // answer groups.
+    // An Equals rule will make all of the following rules with a
+    // matching input, invalid.
+    // A HasNumberOfTermsEqualTo rule will make the following rules of the
+    // same rule type and a matching input, invalid.
+    let seenRules = [];
+
+    for (let i = 0; i < answerGroups.length; i++) {
+      let rules = answerGroups[i].getRulesAsList();
+      for (let j = 0; j < rules.length; j++) {
+        let currentRuleType = <string> rules[j].type;
+        let currentInput = null;
+        var ratio: Ratio = null;
+        if (currentRuleType === 'HasNumberOfTermsEqualTo') {
+          currentInput = <number> rules[j].inputs.y;
+        } else {
+          currentInput = <number[]> rules[j].inputs.x;
+        }
+
+        if (minimumNumberOfTerms > 2) {
+          if (currentRuleType === 'HasNumberOfTermsEqualTo') {
+            if (currentInput < minimumNumberOfTerms) {
+              warningsList.push({
+                type: AppConstants.WARNING_TYPES.ERROR,
+                message: (
+                  'Rule ' + (j + 1) + ' from answer group ' +
+                (i + 1) +
+                ' will never be matched because it has ' +
+                'fewer number of terms than required.')
+              });
+            }
+          } else {
+            ratio = this.rof.fromList(<number[]> currentInput);
+            if (ratio.getNumberOfTerms() < minimumNumberOfTerms) {
+              warningsList.push({
+                type: AppConstants.WARNING_TYPES.ERROR,
+                message: (
+                  'Rule ' + (j + 1) + ' from answer group ' +
+                (i + 1) +
+                ' will never be matched because it has ' +
+                'fewer number of terms than required.')
+              });
+            }
+          }
+        }
+        if (currentRuleType === 'IsEquivalent' && (
+          !this.rof.arrayEquals(
+            ratio.convertToSimplestForm(), currentInput))
+        ) {
```

Comment:
Suggest moving this to a helper to simplify the conditional (see my related comment below).

Reply:
couldn't think of anything much simpler also think keeping the logic here makes more sense as the person using this could read out the thing straight away. Same for below

Reply:
The confusion is that the reader doesn’t need to understand the nuances of the check to reason at the conditional level here. As written, this is hard to read. Consider one alternative:

```typescript
if (!isInExpectedSimplestForm(currentRuleType, ratio, input)) { 
```

You could also split the check into two helper methods. The problem is there are irrelevant details which are adding to the visual complexity of this conditional check. Factoring the logic into helpers puts a named context on what’s being checked and hides the extra details.